src/dataflow/src/source/kinesis.rs

```diff
@@ -37,126 +41,237 @@ lazy_static! {
+        if let Consistency::BringYourOwn(_) = consistency {
+            error!("Kinesis sources do not currently support BYO consistency");
+            None
+        } else {
+            timestamp_data_updates
+                .borrow_mut()
+                .insert(id.clone(), TimestampDataUpdate::RealTime(1));
+            timestamp_metadata_channel
                 .as_ref()
                 .borrow_mut()
-                .push(TimestampMetadataUpdate::StartTimestamping(config.id));
-            Some(config.timestamp_tx)
+                .push(TimestampMetadataUpdate::StartTimestamping(*id));
+            Some(timestamp_metadata_channel)
+        }
+    }
+
+    fn can_close_timestamp(
+        &self,
+        consistency_info: &ConsistencyInfo,
+        pid: &PartitionId,
+        offset: MzOffset,
+    ) -> bool {
+        if !self.is_activated_reader {
+            true
         } else {
-            None
+            // Guaranteed to exist if we receive a message from this partition
+            let last_offset = consistency_info
+                .partition_metadata
+                .get(&pid)
+                .unwrap()
+                .offset;
+            last_offset >= offset
         }
-    } else {
-        None
-    };
+    }
 
-    let mut state = block_on(create_state(connector));
-    let mut last_checked_shards = std::time::Instant::now();
+    fn get_worker_partition_count(&self) -> i32 {
+        if self.is_activated_reader {
+            // Kinesis does not support more than i32 partitions
+            self.shard_set.len().try_into().unwrap()
+        } else {
+            0
+        }
+    }
+
+    fn has_partition(&self, _partition_id: PartitionId) -> bool {
+        self.is_activated_reader
+    }
 
-    let SourceConfig { name, scope, .. } = config;
+    fn ensure_has_partition(&mut self, _consistency_info: &mut ConsistencyInfo, _pid: PartitionId) {
+        //TODO(natacha): do nothing for now, as do not currently use timestamper
+    }
 
-    let (stream, capability) = source(config.id, ts, scope, name.clone(), move |info| {
-        let activator = scope.activator_for(&info.address[..]);
+    fn update_partition_count(
+        &mut self,
+        _consistency_info: &mut ConsistencyInfo,
+        _partition_count: i32,
+    ) {
+        //TODO(natacha): do nothing for now as do not currently use timestamper
+    }
 
-        move |cap, output| {
-            let (client, stream_name, shard_set, shard_queue) = match &mut state {
-                Ok(state) => state,
-                Err(e) => {
-                    error!("failed to create Kinesis state: {:#?}", e);
-                    return SourceStatus::Done;
-                }
-            };
+    fn get_next_message(
+        &mut self,
+        consistency_info: &mut ConsistencyInfo,
+        activator: &Activator,
+    ) -> Result<Option<SourceMessage<Vec<u8>>>, failure::Error> {
+        assert_eq!(self.shard_queue.len(), self.shard_set.len());
```

Comment:
How would this happen? Asking to see if we should give it a chance to recover instead of panicking

Reply:
It should never happen. Just an assert here to make sure that the code doesn't accidentally remove a shard from a set and forget to add it back.